File: lib/providers/whatsapp.ts
/**
 * WhatsApp provider — outbound messages via a BSP (Gupshup / Twilio /
 * Interakt / Meta Cloud API). WhatsApp is the highest-response B2B channel
 * in India/SEA, so it's a first-class outreach + alerting surface here.
 *
 * BSP-agnostic: a single HTTP POST configured by env —
 *   WHATSAPP_API_URL   — BSP send endpoint
 *   WHATSAPP_API_KEY   — bearer token / API key
 *   WHATSAPP_FROM      — sender phone-number id / channel id
 *
 * Mock fallback: when those are unset (or the recipient is the sentinel
 * "mock") we simulate a successful send with a deterministic-ish id, so the
 * whole flow works without a BSP account — matching every other provider.
 *
 * Note on policy: business-INITIATED WhatsApp messages must use a
 * pre-approved template (sendWhatsAppTemplate). Free-form sendWhatsApp is
 * valid only inside the 24h customer-service window (e.g. replies). The
 * caller decides which to use; both mock cleanly.
 */

export function whatsappConfigured(): boolean {
  return Boolean(
    process.env.WHATSAPP_API_URL &&
      process.env.WHATSAPP_API_KEY &&
      process.env.WHATSAPP_FROM,
  )
}

export interface WhatsAppResult {
  messageId: string
  mock: boolean
  error?: string
}

function mockId(): string {
  return `wamid.mock-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`
}

/** Normalise a phone number to digits + leading country code (no +, spaces, dashes). */
export function normalizeWhatsAppNumber(raw: string): string {
  return raw.replace(/[^\d]/g, "")
}

async function postToBsp(payload: Record<string, unknown>): Promise<WhatsAppResult> {
  try {
    const res = await fetch(process.env.WHATSAPP_API_URL as string, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${process.env.WHATSAPP_API_KEY}`,
      },
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(12_000),
    })
    if (!res.ok) {
      return { messageId: "", mock: false, error: `WhatsApp BSP ${res.status}` }
    }
    const json = (await res.json().catch(() => ({}))) as {
      messages?: Array<{ id?: string }>
      id?: string
    }
    const id = json.messages?.[0]?.id ?? json.id ?? `wa-${Date.now().toString(36)}`
    return { messageId: id, mock: false }
  } catch (err) {
    return {
      messageId: "",
      mock: false,
      error: err instanceof Error ? err.message : "WhatsApp send failed",
    }
  }
}

/**
 * Free-form text message (valid inside the 24h reply window — e.g. replies
 * and alert notifications to the user's own number).
 */
export async function sendWhatsApp(opts: {
  to: string
  text: string
}): Promise<WhatsAppResult> {
  const to = normalizeWhatsAppNumber(opts.to)
  if (!whatsappConfigured() || !to || opts.to === "mock") {
    return { messageId: mockId(), mock: true }
  }
  return postToBsp({
    messaging_product: "whatsapp",
    from: process.env.WHATSAPP_FROM,
    to,
    type: "text",
    text: { body: opts.text },
  })
}

/**
 * Template message — required for business-initiated outreach (cold first
 * touch). `params` fill the template's {{1}}, {{2}}, … placeholders.
 */
export async function sendWhatsAppTemplate(opts: {
  to: string
  template: string
  params?: string[]
  languageCode?: string
}): Promise<WhatsAppResult> {
  const to = normalizeWhatsAppNumber(opts.to)
  if (!whatsappConfigured() || !to || opts.to === "mock") {
    return { messageId: mockId(), mock: true }
  }
  return postToBsp({
    messaging_product: "whatsapp",
    from: process.env.WHATSAPP_FROM,
    to,
    type: "template",
    template: {
      name: opts.template,
      language: { code: opts.languageCode ?? "en" },
      components: opts.params?.length
        ? [
            {
              type: "body",
              parameters: opts.params.map((text) => ({ type: "text", text })),
            },
          ]
        : [],
    },
  })
}

File: backend-python/leadgen_backend/providers/whatsapp.py
"""
WhatsApp BSP provider — mirrors lib/providers/whatsapp.ts

Supports free-form messages + pre-approved templates.
Mock-safe when WHATSAPP_API_URL/KEY/FROM are not set.
"""
from __future__ import annotations

import os
from typing import Any

import httpx


def _is_configured() -> bool:
    return bool(
        os.getenv("WHATSAPP_API_URL")
        and os.getenv("WHATSAPP_API_KEY")
        and os.getenv("WHATSAPP_FROM")
    )


def normalize_whatsapp_number(phone: str) -> str:
    """Normalize to E.164 format (digits only, no leading +)."""
    digits = "".join(c for c in phone if c.isdigit())
    if digits.startswith("0"):
        digits = "91" + digits[1:]  # India default
    return digits


async def send_whatsapp(to: str, body: str) -> dict[str, Any]:
    """Send a free-form WhatsApp message."""
    if not _is_configured():
        return {"mock": True, "to": to, "body": body[:50]}

    api_url = os.getenv("WHATSAPP_API_URL", "")
    api_key = os.getenv("WHATSAPP_API_KEY", "")
    from_number = os.getenv("WHATSAPP_FROM", "")

    normalized = normalize_whatsapp_number(to)

    try:
        async with httpx.AsyncClient(timeout=10) as client:
            resp = await client.post(
                f"{api_url.rstrip('/')}/messages",
                json={
                    "messaging_product": "whatsapp",
                    "to": normalized,
                    "type": "text",
                    "text": {"body": body},
                },
                headers={
                    "Authorization": f"Bearer {api_key}",
                    "Content-Type": "application/json",
                },
            )
        resp.raise_for_status()
        return {"ok": True, "to": normalized, **resp.json()}
    except Exception as e:
        return {"ok": False, "error": str(e), "to": normalized}


async def send_whatsapp_template(
    to: str,
    template_name: str,
    language_code: str = "en",
    components: list[dict[str, Any]] | None = None,
) -> dict[str, Any]:
    """Send a pre-approved WhatsApp template message."""
    if not _is_configured():
        return {"mock": True, "to": to, "template": template_name}

    api_url = os.getenv("WHATSAPP_API_URL", "")
    api_key = os.getenv("WHATSAPP_API_KEY", "")

    normalized = normalize_whatsapp_number(to)

    payload: dict[str, Any] = {
        "messaging_product": "whatsapp",
        "to": normalized,
        "type": "template",
        "template": {
            "name": template_name,
            "language": {"code": language_code},
        },
    }
    if components:
        payload["template"]["components"] = components

    try:
        async with httpx.AsyncClient(timeout=10) as client:
            resp = await client.post(
                f"{api_url.rstrip('/')}/messages",
                json=payload,
                headers={
                    "Authorization": f"Bearer {api_key}",
                    "Content-Type": "application/json",
                },
            )
        resp.raise_for_status()
        return {"ok": True, "to": normalized, **resp.json()}
    except Exception as e:
        return {"ok": False, "error": str(e), "to": normalized}


def normalize_whatsapp_payload(payload: dict[str, Any]) -> dict[str, Any]:
    """Normalize incoming WhatsApp webhook payload."""
    messages: list[dict[str, Any]] = []
    statuses: list[dict[str, Any]] = []

    entry_list = payload.get("entry", [])
    for entry in entry_list:
        for change in entry.get("changes", []):
            value = change.get("value", {})
            for msg in value.get("messages", []):
                messages.append({
                    "from": msg.get("from"),
                    "type": msg.get("type"),
                    "body": msg.get("text", {}).get("body") if msg.get("type") == "text" else None,
                    "timestamp": msg.get("timestamp"),
                    "id": msg.get("id"),
                })
            for status in value.get("statuses", []):
                statuses.append({
                    "id": status.get("id"),
                    "status": status.get("status"),
                    "timestamp": status.get("timestamp"),
                    "recipient_id": status.get("recipient_id"),
                })

    return {"messages": messages, "statuses": statuses}
